File: src/forneus_core/exports.py
"""Portable reports, generic over `ToolResult`. Pure serialization, no I/O.

These work for any registered tool: they read `totals`, `rows` and
`notices` from the contract and never import a tool module. A new tool gets
CSV and JSON export without writing a line of serialization code.
"""
import csv
import io
import json

from .contract import ToolResult
from .i18n import DEFAULT_LOCALE, translate, translate_notice

BOM = '﻿'


def report_document(result, locale=DEFAULT_LOCALE):
    """Plain-data snapshot with notices resolved into the viewer's language."""
    document = result.as_document()
    document['locale'] = locale
    document['messages'] = [translate_notice(notice, locale)
                            for notice in result.notices]
    unit_key = result.meta.get('cost_unit_key')
    if unit_key:
        document['meta']['cost_unit_label'] = translate(unit_key, locale=locale)
    return document


def _localize(value, locale):
    """Decimal commas for pt-BR spreadsheets; plain points everywhere else."""
    if isinstance(value, float) and locale == 'pt-BR':
        return str(value).replace('.', ',')
    if isinstance(value, (list, tuple)):
        return '/'.join(map(str, value))
    return value


def format_csv(result, locale=DEFAULT_LOCALE):
    """UTF-8 with BOM and semicolons - what Excel pt-BR opens cleanly.

    One `summary` row carries the totals and the caveats; each `row` line
    carries one entity. Totals stay off the entity rows so summing a column
    never multiplies the report totals by the team size.
    """
    document = report_document(result, locale)
    # Totals are prefixed: a tool may legitimately use the same key for a
    # per-row value and for its total (raw_power, effective_power), and two
    # identical column names would silently collapse into one.
    totals = {f'total_{key}': value for key, value in document['totals'].items()}
    row_columns = sorted({key for row in document['rows'] for key in row})
    columns = ['row_type', *row_columns, *totals, 'messages']

    output = io.StringIO(newline='')
    output.write(BOM)
    writer = csv.DictWriter(output, fieldnames=columns, delimiter=';',
                            extrasaction='ignore')
    writer.writeheader()

    summary = dict(totals)
    summary['row_type'] = 'summary'
    summary['messages'] = ' '.join(document['messages'])

    for row in [summary, *({'row_type': 'row', **row} for row in document['rows'])]:
        writer.writerow({key: _localize(value, locale)
                         for key, value in row.items()})
    return output.getvalue()


def format_json(result, locale=DEFAULT_LOCALE):
    """UTF-8 JSON text: numeric values, provenance and resolved messages."""
    return json.dumps(report_document(result, locale), ensure_ascii=False,
                      indent=2, allow_nan=False) + '\n'

File: web/src/lib/exports.ts
/**
 * CSV and JSON export, ported from `forneus_core.exports`.
 *
 * Generic over the result document, exactly like the Python version: it reads
 * `totals`, `rows` and `notices` and knows nothing about amigatos. A second
 * tool gets both formats for free.
 *
 * The CSV is what Excel pt-BR opens cleanly: UTF-8 with a BOM, semicolons,
 * and decimal commas. Total columns carry a `total_` prefix because a tool may
 * legitimately use one name for both a per-row value and its total —
 * `raw_power` does — and two identical column headers silently collapse.
 */
import type { ToolResultDocument } from './solver/index.ts';

const BOM = '﻿';

export interface ReportOptions {
  locale: string;
  /** Resolves a notice key into text. Injected so this file stays free of i18n. */
  translate: (key: string, params?: Record<string, string | number>) => string;
}

export function reportDocument(result: ToolResultDocument, options: ReportOptions) {
  return {
    ...result,
    locale: options.locale,
    messages: result.notices.map((notice) => options.translate(notice.key, notice.params)),
    meta: {
      ...result.meta,
      cost_unit_label: options.translate(result.meta.cost_unit_key),
    },
  };
}

function csvCell(value: unknown, locale: string): string {
  if (value === undefined || value === null) return '';
  if (Array.isArray(value)) return value.join('/');
  if (typeof value === 'number' && !Number.isInteger(value) && locale.startsWith('pt')) {
    return String(value).replace('.', ',');
  }
  return String(value);
}

function csvEscape(cell: string): string {
  return /[";\n]/.test(cell) ? `"${cell.replace(/"/g, '""')}"` : cell;
}

export function formatCsv(result: ToolResultDocument, options: ReportOptions): string {
  const document = reportDocument(result, options);

  const totals = Object.fromEntries(
    Object.entries(document.totals).map(([key, value]) => [`total_${key}`, value]));
  const rowColumns = [...new Set(document.rows.flatMap((row) => Object.keys(row)))].sort();
  const columns = ['row_type', ...rowColumns, ...Object.keys(totals), 'messages'];

  const lines = [columns.join(';')];

  const summary: Record<string, unknown> = { row_type: 'summary', ...totals };
  summary.messages = document.messages.join(' ');
  lines.push(columns.map((key) =>
    csvEscape(csvCell(summary[key], options.locale))).join(';'));

  for (const row of document.rows) {
    const line: Record<string, unknown> = { row_type: 'row', ...row };
    lines.push(columns.map((key) =>
      csvEscape(csvCell(line[key], options.locale))).join(';'));
  }

  return BOM + lines.join('\r\n') + '\r\n';
}

export function formatJson(result: ToolResultDocument, options: ReportOptions): string {
  return JSON.stringify(reportDocument(result, options), null, 2) + '\n';
}
